SharedPackages/DataBrokerProtectionCore/Sources/DataBrokerProtectionCore/OptOutServices/EmailConfirmationDataService.swift

```diff
@@ -0,0 +1,153 @@
+    private let database: DataBrokerProtectionRepository
+    private let emailServiceV0: EmailServiceProtocol
+    private let emailServiceV1: EmailServiceV1Protocol
+    private let featureFlagger: DBPFeatureFlagging
+    private let pixelHandler: EventMapping<DataBrokerProtectionSharedPixels>?
+
+    public init(database: DataBrokerProtectionRepository,
+                emailServiceV0: EmailServiceProtocol,
+                emailServiceV1: EmailServiceV1Protocol,
+                featureFlagger: DBPFeatureFlagging,
+                pixelHandler: EventMapping<DataBrokerProtectionSharedPixels>?) {
+        self.database = database
+        self.emailServiceV0 = emailServiceV0
+        self.emailServiceV1 = emailServiceV1
+        self.featureFlagger = featureFlagger
+        self.pixelHandler = pixelHandler
+    }
+
+    public func getEmailAndOptionallySaveToDatabase(dataBrokerId: Int64,
+                                                    dataBrokerURL: String,
+                                                    profileQueryId: Int64,
+                                                    extractedProfileId: Int64,
+                                                    attemptId: UUID) async throws -> EmailData {
+        let emailData = try await emailServiceV0.getEmail(dataBrokerURL: dataBrokerURL, attemptId: attemptId)
+
+        if featureFlagger.isEmailConfirmationDecouplingFeatureOn {
+            try database.saveOptOutEmailConfirmation(profileQueryId: profileQueryId,
+                                                     brokerId: dataBrokerId,
+                                                     extractedProfileId: extractedProfileId,
+                                                     generatedEmail: emailData.emailAddress,
+                                                     attemptID: attemptId.uuidString)
+        }
+
+        return emailData
+    }
+
+    public func getConfirmationLink(from email: String,
+                                    numberOfRetries: Int,
+                                    pollingInterval: TimeInterval,
+                                    attemptId: UUID,
+                                    shouldRunNextStep: @escaping () -> Bool) async throws -> URL {
+        try await emailServiceV0.getConfirmationLink(from: email,
+                                                     numberOfRetries: numberOfRetries,
+                                                     pollingInterval: pollingInterval,
+                                                     attemptId: attemptId,
+                                                     shouldRunNextStep: shouldRunNextStep)
+    }
+
+    public func checkForEmailConfirmationData() async throws {
+        let recordsAwaitingLink = try database.fetchOptOutEmailConfirmationsAwaitingLink()
+
+        var itemsToDelete: [EmailDataRequestItemV1] = []
+        for chunk in recordsAwaitingLink.chunks(ofCount: EmailServiceV1.Constants.maxBatchSize) {
```

Comment:
Might be worth adding a comment about the BE limit to explain why this is necessary